hummingbot/connector/budget_checker.py

```diff
@@ -0,0 +1,290 @@
+        charged on top of the order quoted amount; and it assumes that, for sell orders,
+        the fee is deducted from the returns after the sell has completed.
+
+        Example:
+             - Buy Order
+                 The user is submitting an order to buy 100 USDT worth of BTC on an exchange
+                 with a 1% fee. The required collateral will be 101 USDT (i.e. the adjustment term is 1).
+             - Sell Order
+                 The user is submitting an order to sell 1 ETH. The required collateral
+                 is 1 ETH, because the exchange will deduct the 1% fee from the quote asset
+                 returns after the sale has taken place (i.e. the adjustment term is zero).
+
+        Some exchanges can deduct the transaction fee from the buy order amount
+        (i.e. in the buy order example above, the user will only obtain 99 USDT worth of BTC).
+        Yet others can charge the fees for both buy and sell orders in the quote asset.
+        """
+        if order_candidate.order_side == TradeType.BUY:
+            trading_pair = order_candidate.trading_pair
+            price = order_candidate.price
+            amount = order_candidate.amount
+            base, quote = split_hb_trading_pair(trading_pair)
+            fee = self._exchange.get_fee(
+                base, quote, order_candidate.order_type, order_candidate.order_side, amount, price
+            )
+            fee_adjustment = fee.fee_amount_in_quote(trading_pair, price, amount)
+        else:
+            fee_adjustment = Decimal("0")
+        return fee_adjustment
+
+    def _lock_available_collateral(self, order_candidate: OrderCandidate):
+        self._locked_collateral[order_candidate.collateral_token] += order_candidate.collateral_amount
+
+    def _adjust_amounts(
+        self, order_candidate: OrderCandidate, available_collateral: Decimal
+    ) -> OrderCandidate:
+        if order_candidate.order_side == TradeType.BUY:
+            trading_pair = order_candidate.trading_pair
+            base, quote = split_hb_trading_pair(trading_pair)
+            fee = self._exchange.get_fee(
+                base,
+                quote,
+                order_candidate.order_type,
+                order_candidate.order_side,
+                order_candidate.amount,
+                order_candidate.price,
+            )
+            adjusted_amount = fee.order_amount_from_quote_with_fee(
+                trading_pair, order_candidate.price, available_collateral
+            )
+            adjusted_amount = self._exchange.quantize_order_amount(trading_pair, adjusted_amount)
+        else:
+            adjusted_amount = available_collateral / order_candidate.price
+
+        adjusted_candidate = copy(order_candidate)
+        adjusted_candidate.amount = adjusted_amount
+        adjusted_candidate.collateral_amount = available_collateral
+
+        return adjusted_candidate
+
+
+class PerpetualBudgetChecker(BudgetChecker):
```

Comment:
It might be better to move the PerpetualBudgetChecker to its own module, and place it inside the perpetual connectors folder.